File: packages/ruby/lib/nurse_andrea/environment_detector.rb
module NurseAndrea
  class EnvironmentDetector
    SUPPORTED = %w[production staging development].freeze

    class << self
      def detect
        raw = ENV["RAILS_ENV"] || ENV["RACK_ENV"]
        return "production" if raw.nil? || raw.empty?

        return raw if SUPPORTED.include?(raw)

        warn_unsupported(raw)
        "production"
      end

      def reset_warning!
        @warned = false
      end

      private

      def warn_unsupported(value)
        return if @warned

        @warned = true
        $stderr.puts(
          "[NurseAndrea] Detected environment '#{value}' is not in the " \
          "supported set #{SUPPORTED.inspect}. Falling back to 'production'."
        )
      end
    end
  end
end

File: packages/python/src/nurse_andrea/environment_detector.py
import os
import sys

SUPPORTED_ENVIRONMENTS = ("production", "staging", "development")

_warned = False


def detect_environment() -> str:
    raw = (
        os.environ.get("PYTHON_ENV")
        or os.environ.get("ENV")
        or os.environ.get("APP_ENV")
    )
    if not raw:
        return "production"

    if raw in SUPPORTED_ENVIRONMENTS:
        return raw

    _warn_once(raw)
    return "production"


def _warn_once(value: str) -> None:
    global _warned
    if _warned:
        return
    _warned = True
    sys.stderr.write(
        f"[NurseAndrea] Detected environment '{value}' is not in the "
        f"supported set {list(SUPPORTED_ENVIRONMENTS)}. "
        "Falling back to 'production'.\n"
    )


def _reset_warning() -> None:
    global _warned
    _warned = False
